### PR TITLE
unified installer

### LINKED ISSUE #635: Installing On windows 10 Pip error
```
Installing collected packages: EXIFRead, urllib3, certifi, idna, chardet, six, asn1crypto, pycparser, cffi, cryptography, pyOpenSSL, requests, setuptools, MarkupSafe, Jinja2, PyYAML, argh, pathtools, watchdog, click, itsdangerous, Werkzeug, Flask, pip, mistune, inifile, pytz, Babel, Lektor
  Running setup.py install for EXIFRead ... done
  Running setup.py install for pycparser ... done
  Found existing installation: setuptools 28.8.0
    Uninstalling setuptools-28.8.0:
      Successfully uninstalled setuptools-28.8.0
  Running setup.py install for pathtools ... done
  Running setup.py install for watchdog ... done
  Found existing installation: pip 9.0.1
    Uninstalling pip-9.0.1:
      Successfully uninstalled pip-9.0.1
Exception:
Traceback (most recent call last):
  File "C:\Python36\lib\shutil.py", line 387, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\LenovoPC\\AppData\\Local\\Temp\\pip-okdblfrr-uninstall\\users\\lenovopc\\appdata\\local\\lektor-cli\\lib\\scripts\\pip.exe'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\basecommand.py", line 215, in main
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\commands\install.py", line 342, in run
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\req\req_set.py", line 795, in install
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\req\req_install.py", line 767, in commit_uninstall
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\req\req_uninstall.py", line 142, in commit
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\_vendor\retrying.py", line 49, in wrapped_f
    return Retrying(*dargs, **dkw).call(f, *args, **kw)
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\_vendor\retrying.py", line 212, in call
    raise attempt.get()
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\_vendor\retrying.py", line 247, in get
    six.reraise(self.value[0], self.value[1], self.value[2])
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\_vendor\six.py", line 686, in reraise
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\_vendor\retrying.py", line 200, in call
    attempt = Attempt(fn(*args, **kwargs), attempt_number, False)
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\utils\__init__.py", line 102, in rmtree
  File "C:\Python36\lib\shutil.py", line 494, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "C:\Python36\lib\shutil.py", line 384, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "C:\Python36\lib\shutil.py", line 384, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "C:\Python36\lib\shutil.py", line 384, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  [Previous line repeated 3 more times]
  File "C:\Python36\lib\shutil.py", line 389, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "c:\users\lenovopc\appdata\local\lektor-cli\lib\lib\site-packages\pip\utils\__init__.py", line 114, in rmtree_errorhandler
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\LenovoPC\\AppData\\Local\\Temp\\pip-okdblfrr-uninstall\\users\\lenovopc\\appdata\\local\\lektor-cli\\lib\\scripts\\pip.exe'

All done!
```

### LINKED ISSUE #720: Installer broken with virtualenv 20.x
I tried installing Lektor on Linux via the installer script at `https://www.getlektor.com/install.sh`. This resulted in the following error messages:
```
/usr/bin/python: can't open file './src/virtualenv.py': [Errno 2] No such file or directory
/usr/bin/python: can't open file './virtualenv.py': [Errno 2] No such file or directory
```
It seems the internal structure of the `virtualenv` package changed again. If I modify the installer script to force an older version of virtualenv, it works again:
```py
VENV_URL = "https://pypi.org/pypi/virtualenv/16.7.9/json"
```

### PR DESCRIPTION
**Pull request by xlotlu**
~~This fixes #720, but does not fix #635 (yet, more on that later).~~
This fixes #720 and #635 (but uncovers a new windows issue, more on that later).

So, the big change:
 - fix the great annoyance that was separate installers for *nix and windows,
 - while getting rid of the python embedded in sh / ps,
 - and moving the code over to the main project, out of the website.

What I intend to do:
 - import the old install.sh and install.ps with all their commit history;
 - update the *nix docs to download and run the python file directly (using a .sh was pointless);
 - figure out a way to handle windows. The old instructions were downloading the .ps using ps scripting anyway, shouldn't be too hard.
 - for now copy the installer(s) manually over to the website assets; set up an automated job later on.

Now, I need help with the following:
 - code review. Not so much the logic (which I think I reproduced accurately from the old code), but rather the style and structure. I tried to make it elegant and logical, but I'm not entirely happy with it.
 - a crazy spectacular issue with windows 10. I'll describe it in a new message.

### CODE REVIEW

File: bin/install.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        paths = [
+            item for item in os.environ['PATH'].split(':')
+            if not item.endswith('/sbin') and os.access(item, os.W_OK)
+        ]
+
+        if not paths:
+            fail('None of the items in $PATH are writable. Run with '
+                 'sudo or add a $PATH item that you have access to.')
+
+        def _sorter(path):
```
Comment (runfalk) [inline bin/install.py]: Maybe a comment here that we prioritize bin paths based on position in KNOWN_BIN_PATHS? And that this ensures that we always do a system install when possible. At least that's my understanding of this.

File: bin/install.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        if os.path.lexists(self.symlink_path):
+            os.remove(self.symlink_path)
+        if os.path.exists(self.lib_dir):
+            shutil.rmtree(self.lib_dir, onerror=self.deletion_error)
+
+    def get_pip(self):
+        return os.path.join(self.lib_dir, 'bin', 'pip')
+
+    def install_lektor(self):
+        import pdb
```
Comment (runfalk) [inline bin/install.py]: Forgotten pdb?

**Comment by xlotlu** [comment]
So, under Win 10 the installer works only when installing outside AppData\Local.

When writing under AppData\Local, this spectacular thing happens:

```python
In [1]: import os

In [2]: D = os.environ['LocalAppData']

In [3]: print(D)
C:\Users\admin\AppData\Local

In [4]: os.getlogin()
Out[4]: 'admin'

In [5]: os.mkdir(os.path.join(D, "testing123"))

In [6]: os.stat(os.path.join(D, "testing123"))
Out[6]: os.stat_result(st_mode=16895, st_ino=6473924464382364, st_dev=821167655, st_nlink=1, st_uid=0, st_gid=0, st_size=0, st_atime=1584711422, st_mtime=1584711422, st_ctime=1584711422)

In [7]:
Do you really want to exit ([y]/n)?
```

Everything seems in order. Let's check with the windows tools:

```
C:\Users\admin>dir C:\Users\admin\AppData\Local\testing123
 Volume in drive C has no label.
 Volume Serial Number is 30F2-0627

 Directory of C:\Users\admin\AppData\Local

File Not Found

```

Now that's strange. Let's go back to Python:

```python
In [1]: import os

In [2]: os.stat("C://Users/admin/AppData/Local/testing123")
Out[2]: os.stat_result(st_mode=16895, st_ino=6473924464382364, st_dev=821167655, st_nlink=1, st_uid=0, st_gid=0, st_size=0, st_atime=1584711422, st_mtime=1584711422, st_ctime=1584711422)

In [3]:
```

So, only Python sees those files. They are there, but cmd and explorer don't show them, not even in administrator mode. :thinking: 

In practice it leads to this:

```python
In [1]: import os

In [2]: import shutil

In [3]: D = os.path.join(os.environ['LocalAppData'], "testing123")

In [4]: D
Out[4]: 'C:\\Users\\admin\\AppData\\Local\\testing123'

In [5]: shutil.rmtree(D)

In [6]: from venv import EnvBuilder

In [7]: env = EnvBuilder(with_pip=False)

In [8]: env.create(D)

In [9]: os.listdir(D)
Out[9]: ['Include', 'Lib', 'pyvenv.cfg', 'Scripts']

In [10]: os.listdir(D + "/Scripts")
Out[10]:
['activate',
 'activate.bat',
 'Activate.ps1',
 'deactivate.bat',
 'python.exe',
 'pythonw.exe']

In [11]: from subprocess import call

In [12]: call(os.path.join(D, "Scripts", "python.exe"))
No pyvenv.cfg file
Out[12]: 106

In [13]: from subprocess import check_output

In [14]: check_output(os.path.join(D, "Scripts", "python.exe"))
No pyvenv.cfg file
---------------------------------------------------------------------------
CalledProcessError                        Traceback (most recent call last)
<ipython-input-15-327140e0661e> in <module>
----> 1 check_output(os.path.join(D, "Scripts", "python.exe"))

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.752.0_x64__qbz5n2kfra8p0\lib\subprocess.py in check_output(timeout, *popenargs, **kwargs)
    409         kwargs['input'] = '' if kwargs.get('universal_newlines', False) else b''
    410
--> 411     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
    412                **kwargs).stdout
    413

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.752.0_x64__qbz5n2kfra8p0\lib\subprocess.py in run(input, capture_output, timeout, check, *popenargs, **kwargs)
    510         retcode = process.poll()
    511         if check and retcode:
--> 512             raise CalledProcessError(retcode, process.args,
    513                                      output=stdout, stderr=stderr)
    514     return CompletedProcess(process.args, retcode, stdout, stderr)

CalledProcessError: Command 'C:\Users\admin\AppData\Local\testing123\Scripts\python.exe' returned non-zero exit status 106.

In [15]:
```

Which is exactly what happens when one tries bootstrapping the venv with pip included:

```
C:\Users\admin>python -m venv C:\Users\admin\AppData\Local\moooo
Error: Command '['C:\\Users\\admin\\AppData\\Local\\moooo\\Scripts\\python.exe', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 106.
```

Contrast this with:

```
C:\Users\admin>python -m venv C:\wow\waw\wee\woo

C:\Users\admin>dir C:\wow\waw\wee\woo
 Volume in drive C has no label.
 Volume Serial Number is 30F2-0627

 Directory of C:\wow\waw\wee\woo

03/20/2020  04:16 PM    <DIR>          .
03/20/2020  04:16 PM    <DIR>          ..
03/20/2020  04:16 PM    <DIR>          Include
03/20/2020  04:16 PM    <DIR>          Lib
03/20/2020  04:16 PM               164 pyvenv.cfg
03/20/2020  04:16 PM    <DIR>          Scripts
               1 File(s)            164 bytes
               5 Dir(s)  50,396,495,872 bytes free
```

So. I'm at a loss. This same thing works under Win7 just fine.

File: bin/install.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+def fail(message):
+    print('Error: %s' % message)
+    sys.exit(1)
+
+def multiprint(*lines):
+    for line in lines:
+        print(line)
+
+
+class Installer(object):
```
Comment (runfalk) [inline bin/install.py]: I see what you are going for here, and I assume this is also what you're not completely satisfied with?

To me it looks simpler to declare a separate `install_posix` and `install_windows`. I get that we risk that they diverge but they should be a lot easier to troubleshoot. The bits that are used by both should probably be moved to separate functions.

**Review by runfalk — CHANGES_REQUESTED** [review]
Nice work, I also like the idea of moving it to the Lektor code itself.

I also think we should add a CI job that tests this with `LEKTOR_SILENT`. It should be fairly straight forward, at least for Linux and macOS with 2.7 and 3.8.

**Comment by runfalk** [comment]
@xlotlu maybe it's related to this? https://blogs.windows.com/windowsdeveloper/2016/05/24/sharing-your-local-app-data/

Seems like Windows 10 can isolate applications from each other when it comes to AppData.

If so, maybe AppData is no longer the correct place to install user programs in.

File: bin/install.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+if not _is_win:
+    sys.stdin = open('/dev/tty', 'r')
+
+if sys.version_info[0] == 2:
+    input = raw_input
+
+prompt = os.environ.get('LEKTOR_SILENT') is None
+
+def get_confirmation():
+    if prompt is False:
```
[the remark this answers is not in the document]
Comment (runfalk) [reply bin/install.py]: True, I missed that. Maybe we should do `bool(os.environ.get(...))` then?

[the remark this answers is not in the document]
Comment (runfalk) [reply bin/install.py]: I don't expect LEKTOR_SILENT= pythonscript to run silent, but maybe I'm naive here.

File: bin/install.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+            print('')
+            return
+
+        if user_input == 'n':
+            print('')
+            print('Aborted!')
+            sys.exit()
+
+def fail(message):
+    print('Error: %s' % message)
```
[the remark this answers is not in the document]
Comment (runfalk) [reply bin/install.py]: Technically we don't have to, but yeah I get your point.

File: bin/install.py (hunk 3 above)
Comment (runfalk) [reply bin/install.py]: The thing that bothers me in this solution is that the methods manage their own state anyway. `compute_location` will assign to different things depending on in which class it resides, and subsequent methods are very dependent on that. As someone who didn't write the code I need to bounce back and forth between `Installer.__init__` and the implementation classes since they have lost all context of ordering.

This is why I favor separate functions, but then again I haven't looked at this as deeply as you have. I trust your judgement if you think this is better.

File: bin/install.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+def get_confirmation():
+    if prompt is False:
+        return
+
+    while True:
+        user_input = input('Continue? [Yn] ').lower().strip()
+
+        if user_input in ('', 'y'):
+            print('')
```
[the remark this answers is not in the document]
Comment (runfalk) [reply bin/install.py]: I think it's fine to rip it out and use multiple print calls if you want to.

File: bin/install.py (hunk 3 above)
Comment (xlotlu) [reply bin/install.py]: > The thing that bothers me in this solution is that the methods manage their own state anyway. `compute_location` will assign to different things depending on in which class it resides, and subsequent methods are very dependent on that. As someone who didn't write the code I need to bounce back and forth between `Installer.__init__` and the implementation classes since they have lost all context of ordering.

Yes, I know...

**Comment by xlotlu** [comment]
> @xlotlu maybe it's related to this? https://blogs.windows.com/windowsdeveloper/2016/05/24/sharing-your-local-app-data/
> 
> Seems like Windows 10 can isolate applications from each other when it comes to AppData.

That must be related. It's Python's isolation that does this in fact: I found my files under 

```
C:\Users\admin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\LocalCache\Local\lektor-cli\lib
```
Crazy windows must be virtualizing filesystem access. So, this being the Python version installed from the MS Store, it's a sandboxed application. I uninstalled it and set up the standalone installer instead:

```
λ python install.py

Welcome to Lektor

This script will install Lektor on your computer.

Installing at:
  C:\Users\admin\AppData\Local\lektor-cli

Continue? [Yn]

Collecting Lektor
  Downloading https://files.pythonhosted.org/packages/ ......
...
etc.
```

:)

Now, what's the right way to go about this I have no idea. But at least we know it works.

**Comment by runfalk** [comment]
Does it work even with the Windows store install? Or will the user have to download the normal Python installer?

**Comment by xlotlu** [comment]
> Does it work even with the Windows store install? Or will the user have to download the normal Python installer?

It works if you install anywhere outside of AppData, as that is virtualised for store apps. Maybe the behaviour could  be circumvented (if someone feels like doing lots of reading), or we could choose an alternate install location. Or we could just leave it as is, make a note in the docs how we require the "normal" installer,  and create an issue to deal with it later. (This one's my personal favourite.)

**Comment by runfalk** [comment]
Gave the function-based implementation a quick go (completely untested). I just wanted to compare the readability and structure of it. Right now it's 350 lines vs. 400 in the class based solution.

```python
#!/usr/bin/env python
from __future__ import print_function

import math
import os
import shutil
import sys
import tempfile
from contextlib import contextmanager
from subprocess import call

try:
    from urllib.request import urlretrieve
except ImportError:
    from urllib import urlretrieve


if sys.version_info.major == 2:
    input = raw_input


APP_NAME = "lektor"
VIRTUALENV_URL = "https://bootstrap.pypa.io/virtualenv.pyz"

_IS_WIN = sys.platform == "win32"
if not _IS_WIN:
    sys.stdin = open("/dev/tty", "r")

_PROMPT = os.environ.get("LEKTOR_SILENT", "") in ("", "0", "off", "false")


def continue_prompt():
    if _PROMPT is False:
        return

    while True:
        user_input = input("Continue? [Yn] ").lower().strip()

        if user_input in ("", "y"):
            print()
            return

        if user_input == "n":
            print()
            print("Aborted!")
            sys.exit()


def fail(message):
    print("Error: %s" % message, file=sys.stderr)
    sys.exit(1)


def multiprint(*lines, **kwargs):
    for line in lines:
        print(line, **kwargs)


class Progress(object):
    "A context manager to be used as a urlretrieve reporthook."

    def __init__(self):
        self.started = False

    def progress(self, count, bsize, total):
        size = count * bsize

        if size > total:
            progress = 100
        else:
            progress = math.floor(100 * size / total)

        out = sys.stdout
        if self.started:
            out.write("\b" * 4)

        out.write("{:3d}%".format(progress))
        out.flush()

        self.started = True

    def finish(self):
        sys.stdout.write("\n")

    def __enter__(self):
        return self.progress

    def __exit__(self, exc_type, exc_value, traceback):
        self.finish()


@contextmanager
def fetch_temp_virtualenv():
    fname = os.path.basename(VIRTUALENV_URL)
    root, ext = os.path.splitext(fname)

    zipapp = tempfile.mktemp(prefix=root + "-", suffix=ext)

    with Progress() as hook:
        sys.stdout.write("Downloading virtualenv: ")
        urlretrieve(VIRTUALENV_URL, zipapp, reporthook=hook)

    yield zipapp
    os.unlink(zipapp)


def create_virtualenv(target_dir):
    """
    Tries to create a virtualenv by using the built-in `venv` module,
    or using the `virtualenv` executable if present, or falling back
    to downloading the official zipapp.
    """

    created = False

    # Prioritize Python 3 venv module
    try:
        from venv import EnvBuilder
    except ImportError:
        pass
    else:
        try:
            # TODO: design decision needed:
            # on Debian and Ubuntu systems Python is missing `ensurepip`,
            # prompting the user to install `python3-venv` instead.
            # we could do the same, or go the download route...
            import ensurepip
        except ImportError:
            pass
        else:
            venv = EnvBuilder(with_pip=True, symlinks=False if _IS_WIN else True)
            venv.create(target_dir)
            return

    # Fall back on virtualenv if it's installed
    try:
        from shutil import which
    except ImportError:
        from distutils.spawn import find_executable as which

    venv_exec = which("virtualenv")
    if venv_exec:
        retval = call([venv_exec, "-p", sys.executable, target_dir])
        if retval:
            sys.exit(1)
        return

    # Resort to download virtualenv as a zipapp
    with fetch_temp_virtualenv() as zipapp:
        retval = call([sys.executable, zipapp, target_dir])
        if retval:
            sys.exit(1)


def delete_files_recursive(*paths):
    def abort_on_error(func, path, excinfo):
        multiprint(
            "Problem deleting {}".format(path),
            "Please try and delete {} manually".format(path),
            "Aborted!",
            file=sys.stderr,
        )
        sys.exit(1)
    for path in paths:
        if os.path.isdir(path):
            shutil.rmtree(lib_dir, onerror=abort_on_error)
        else:
            os.remove(path)


def posix_get_bin_and_lib_dir():
    home_dir = os.environ.get("HOME")
    known_bin_paths = [
        "/usr/local/bin",
        "/opt/local/bin",
        os.path.join(home_dir, ".bin"),
        os.path.join(home_dir, ".local", "bin"),
    ]

    # look for writable directories in the user's $PATH
    # (that are not sbin)
    paths = [
        item
        for item in os.environ["PATH"].split(":")
        if not item.endswith("/sbin") and os.access(item, os.W_OK)
    ]

    if not paths:
        fail(
            "None of the items in $PATH are writable. Run with "
            "sudo or add a $PATH item that you have access to."
        )

    # ... and prioritize them according to known_bin_paths.
    # this makes sure we perform a system install when possible.
    def _sorter(path):
        try:
            return known_bin_paths.index(path)
        except ValueError:
            return float("inf")

    paths.sort(key=_sorter)

    for path in paths:
        if path.startswith(_HOME):
            return (path, os.path.join(_HOME, ".local", "lib", APP_NAME))

        if path.endswith("/bin"):
            parent = os.path.dirname(path)
            return (path, os.path.join(parent, "lib", APP_NAME))

    fail("Could not determine installation location for Lektor.")


def install_posix():
    bin_dir, lib_dir = posix_get_bin_and_lib_dir()
    symlink_path = os.path.join(bin_dir, APP_NAME)

    multiprint(
        "Lektor will be installed at:",
        "  bin path: %s" % symlink_path,
        "  lib path: %s" % lib_dir,
        "",
    )
    if os.path.exists(lib_dir) or os.path.exists(symlink_path):
        multiprint(
            "An existing installation was detected. If you continue the existing installation will be removed.",
            "",
        )

    continue_prompt()

    # Remove existing installations before continuing
    delete_files_recursive(symlink_path, lib_dir)

    create_virtualenv(lib_dir)

    pip_path = os.path.join(lib_dir, "bin", "pip")
    call([pip_path, "install", "--upgrade", "Lektor"])

    os.symlink(os.path.join(lib_dir, "bin", "lektor"), symlink_path)


def windows_add_to_path(location):
    try:
        from winreg import (
            OpenKey,
            CloseKey,
            QueryValueEx,
            SetValueEx,
            HKEY_CURRENT_USER,
            KEY_ALL_ACCESS,
            REG_EXPAND_SZ,
        )
    except ImportError:  # py2
        from _winreg import (
            OpenKey,
            CloseKey,
            QueryValueEx,
            SetValueEx,
            HKEY_CURRENT_USER,
            KEY_ALL_ACCESS,
            REG_EXPAND_SZ,
        )
    import ctypes
    from ctypes.wintypes import HWND, UINT, WPARAM, LPARAM, LPVOID

    HWND_BROADCAST = 0xFFFF
    WM_SETTINGCHANGE = 0x1A

    reg_key = OpenKey(HKEY_CURRENT_USER, "Environment", 0, KEY_ALL_ACCESS)

    try:
        path_value, _ = QueryValueEx(reg_key, "Path")
    except WindowsError:
        path_value = ""

    paths = path_value.split(";")
    if location not in paths:
        paths.append(location)
        path_value = ";".join(paths)
        SetValueEx(reg_key, "Path", 0, REG_EXPAND_SZ, path_value)

    SendMessage = ctypes.windll.user32.SendMessageW
    SendMessage.argtypes = HWND, UINT, WPARAM, LPVOID
    SendMessage.restype = LPARAM
    SendMessage(HWND_BROADCAST, WM_SETTINGCHANGE, 0, u"Environment")


def install_windows():
    # lektor-cli instead of lektor for backwards-compatibility with previous installer
    install_dir = os.path.join(os.environ["LocalAppData"], "lektor-cli")
    lib_dir = os.path.join(install_dir, "lib")
    link_path = os.path.join(install_dir, "lektor.cmd")

    multiprint(
        "Lektor will be installed at:",
        "  path: %s" % install_dir, "",
    )
    if os.path.exists(lib_dir) or os.path.exists(symlink_path):
        multiprint(
            "An existing installation was detected. If you continue the existing installation will be removed.",
            "",
        )

    continue_prompt()

    # Remove existing installations before continuing
    delete_files_recursive(install_dir)

    pip_path = os.path.join(lib_dir, "Scripts", "pip.exe")
    call([pip_path, "install", "--upgrade", "Lektor"])

    # Create launch command (bat file that is added to path)
    exe_path = os.path.join(lib_dir, "Scripts", "lektor.exe")
    with open(link_path, "w") as link_file:
        link_file.write("@echo off\n")
        link_file.write('"{}" %*'.format(exe_path))

    windows_add_to_path(install_dir)


if __name__ == "__main__":
    multiprint(
        "",
        "Welcome to Lektor",
        "",
        "This script will install Lektor on your computer.",
        "",
    )

    if _IS_WIN:
        install_windows()
    else:
        install_posix()

    multiprint(
        "", "All done!",
    )
```

**Comment by xlotlu** [comment]
> Gave the function-based implementation a quick go (completely untested). I just wanted to compare the readability and structure of it.

I like some of the logic fixups you made, the single prompt, I'm somewhat partial to the contextmanager, I like some of the renaming, not all. I really dislike black messing with my beautiful handcrafted indentation. :)

I don't like the long main install_* functions, the mashup is hard to follow, but that's an easy fix to regroup the code into the previous functions. But the whole structure makes me dizzy. I mean, I'd even go so far so as to make them all `@staticmethod`s just to have them grouped somehow. :)

The only thing I really didn't like in my version was that one method setting state while others would use it, but I still like it better. On the other hand, I'm biased.

(I guess that ugliness could be improved a bit by setting those attributes in `__init__` instead.)

@goanpeca, @nixjdm, thoughts?

**Comment by runfalk** [comment]
I don't really want to block this (since the installer is currently broken), especially since we might end up rewriting it as a pipx installer or similar anyway down the road. I say borrow whatever you like from my suggestions and finish your implementation.

**Comment by xlotlu** [comment]
@runfalk please apply your changes as a commit

**Comment by xlotlu** [comment]
@runfalk so this is it. I'll move to importing the old history next.

There's only the question of how to get this file over to the website.

**Comment by runfalk** [comment]
I think it looks good.

We probably want to deploy this to lektor-website initially since auto deployment is configured there. However, the build currently breaks on lektor-atom which is owned by @nixjdm so we can't fix it without him.

Alternatively we can point the curl URL to the raw file in this repo, which may be a bit better? Still dependent on lektor-atom getting fixed though.

**Review by runfalk — APPROVED** [review]

**Comment by xlotlu** [comment]
Closing this in favour of https://github.com/lektor/lektor-website/pull/298.

Will port it back here at some point in the future.
